src/interface/swarm.cpp

```diff
@@ -861,58 +995,86 @@ void Swarm::UnloadBuffers_() {
           }
           for (int i = 0; i < int_vars_size; i++) {
             vint(i, sid) = static_cast<int>(
-                bdvar.recv[nid](bid * particle_size + real_vars_size + i));
-          }
-
-          double &x = vreal(ix, sid);
-          double &y = vreal(iy, sid);
-          double &z = vreal(iz, sid);
-          // TODO(BRR) Don't hardcode periodic boundary conditions
-          if (x < swarm_d.x_min_global_) {
-            x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);
-          }
-          if (x > swarm_d.x_max_global_) {
-            x = swarm_d.x_min_global_ + (x - swarm_d.x_max_global_);
-          }
-          if (y < swarm_d.y_min_global_) {
-            y = swarm_d.y_max_global_ - (swarm_d.y_min_global_ - y);
-          }
-          if (y > swarm_d.y_max_global_) {
-            y = swarm_d.y_min_global_ + (y - swarm_d.y_max_global_);
-          }
-          if (z < swarm_d.z_min_global_) {
-            z = swarm_d.z_max_global_ - (swarm_d.z_min_global_ - z);
-          }
-          if (z > swarm_d.z_max_global_) {
-            z = swarm_d.z_min_global_ + (z - swarm_d.z_max_global_);
+                bdvar.recv[nid](real_vars_size + bid * particle_size + i));
           }
         });
+
+    ApplyBoundaries_(total_received_particles_, new_indices);
   }
 }
 
+void Swarm::ApplyBoundaries_(const int nparticles, ParArrayND<int> indices) {
+  auto pmb = GetBlockPointer();
+  auto &x = Get<Real>("x").Get();
+  auto &y = Get<Real>("y").Get();
+  auto &z = Get<Real>("z").Get();
+  auto swarm_d = GetDeviceContext();
+  auto bcs = this->pbounds;
+
+  pmb->par_for(
+      "Swarm::ApplyBoundaries", 0, nparticles - 1, KOKKOS_LAMBDA(const int n) {
+        const int sid = indices(n);
+        for (int l = 0; l < 6; l++) {
+          bcs.bounds[l]->Apply(sid, x(sid), y(sid), z(sid), swarm_d);
+        }
+      });
+
+  RemoveMarkedParticles();
+}
+
 bool Swarm::Receive(BoundaryCommSubset phase) {
-  // Ensure all local deep copies marked BoundaryStatus::completed are actually received
-  GetBlockPointer()->exec_space.fence();
   auto pmb = GetBlockPointer();
+  const int nneighbor = pmb->pbval->nneighbor;
 
-  // Populate buffers
-  vbswarm->Receive(phase);
+  if (nneighbor == 0) {
+    // Do nothing; no boundaries to receive
+    return true;
+  } else {
+    // Ensure all local deep copies marked BoundaryStatus::completed are actually received
+    GetBlockPointer()->exec_space.fence();
+    auto pmb = GetBlockPointer();
```

Comment:
```suggestion
    pmb->exec_space.fence();
```
`pmb` is already defined outside the conditional, isn't it?

Reply:
Fixed